### PR TITLE
spack.yaml: depend on generic tracers Spack components

### PR DESCRIPTION
**Pull request by harshula**
Opening this PR to create a pre-release build of ACCESS-OM2 with Spack-built generic tracers components for testing.

### CODE REVIEW

**Comment by harshula** [comment]
!bump major

**Comment by CodeGat** [comment]
Token has expired. Will regenerate one tomorrow...considering making it have a longer lifespan 😅

**Comment by aidanheerdegen** [comment]
> Token has expired. Will regenerate one tomorrow...considering making it have a longer lifespan 😅

Would be good to have a mechanism to remind us when things like this are expiring. Can we generate tokens through an API and set through an API?

File: spack.yaml [hunk 1]
```diff
@@ -53,8 +73,8 @@ spack:
               'SPACK_{name}_ROOT': '{prefix}'
         projections:
           all: '{name}/{version}'
-          access-om2: '{name}/2024.03.0'
+          access-om2: '{name}/2024.10.0'
           cice5: '{name}/2023.10.19'
-          mom5: '{name}/2023.11.09'
+          mom5: '{name}/development'
```
Comment (aidanheerdegen) [inline spack.yaml:78]: This made me wonder, would we have namespace clash issues with modules in `prerelease` if two PRs tried to create modules using the same tag or name?

ping @CodeGat for visibility

**Review by aidanheerdegen — COMMENTED** [review]
Just a question at this point.

This is fine for testing, but just flagging we have some fork management discussions/decisons to make for those new components before merging.

File: spack.yaml (hunk 1 above)
Comment (CodeGat) [reply spack.yaml:78]: I suspect so, now that you mention it...!

Comment (harshula) [reply spack.yaml:81]: Shall we do `development_2024.10.08` or `dev_2024.10.08`?

Comment (aidanheerdegen) [reply spack.yaml:81]: `dev_2024.10.08` is sufficiently descriptive and a little less clunky.

Comment (aidanheerdegen) [reply spack.yaml:81]: For the future: could we isolate modules into separate namespaces, use the PR name as well?

**Comment by CodeGat** [comment]
Note: the Prerelease succeeded (and is usable), the workflow failed to collect some metadata as the updates to `spack-config` move directories around. Will link an issue soon.

File: spack.yaml (hunk 1 above)
Comment (CodeGat) [reply spack.yaml:81]: How do we do that? Just change the name of the module, or something fancy with modules that I don't know?

**Comment by harshula** [comment]
Hi @chrisb13 , @anton-seaice , @dougiesquire , This is ready for testing.

**Comment by dougiesquire** [comment]
> Hi @chrisb13 , @anton-seaice , @dougiesquire , This is ready for testing.

If anyone wants to try run this, they can use [this config](https://github.com/ACCESS-NRI/access-om2-configs/tree/dougiesquire/1deg_jra55_ryf_wombatlite). You'll obviously need to modify the `config.yaml` to use the prerelease module.

**Comment by chrisb13** [comment]
Following @dougiesquire suggestion, we've ran a test using this [branch](https://github.com/ACCESS-NRI/access-om2-configs/tree/dougiesquire/1deg_jra55_ryf_wombatlite) (commit `12761ac56fc9c7fa2c80db17f1b0d8c337724ea2`). With small modifications to the [config.yaml](https://github.com/user-attachments/files/17337217/config.yaml.txt) (thanks @jo-basevi!).

The test ran successfully for five years and produced output from ACCESS-OM2 with WOMBAT legacy in generic tracers.

@pearseb also checked the output. NPP screenshot below.

image001

**Comment by aidanheerdegen** [comment]
> With small modifications to the [config.yaml](https://github.com/user-attachments/files/17337217/config.yaml.txt) (thanks @jo-basevi!).

Can we get those modifications into this PR?

> The test ran successfully for five years and produced output from ACCESS-OM2 with WOMBAT legacy in generic tracers.

Nice. Was this up for debate?

**Comment by chrisb13** [comment]
> Can we get those modifications into this PR?

The modifications were small but also made to the config repo' [(here)](https://github.com/ACCESS-NRI/access-om2-configs/tree/dougiesquire/1deg_jra55_ryf_wombatlite) so not sure how they'd be incorporated into this PR..?

**Comment by aidanheerdegen** [comment]
Oh sorry, I misspoke. As you quite rightly point out the [config repo](https://github.com/ACCESS-NRI/access-om2-configs/tree/dougiesquire/1deg_jra55_ryf_wombatlite) isn't part of this PR, so ignore me. But you can push changes to that branch, or create a new branch with those changes, to make it easier to test.

**Comment by chrisb13** [comment]
> But you can push changes to that branch, or create a new branch with those changes, to make it easier to test.

Ok as discussed and above, now on a new branch on my own repo', here:
https://github.com/chrisb13/access-om2-configs/commit/e41c923a600dbb309d3286803c36976a61f73501
Do I put in a pull request? (Does it need to get onto the https://github.com/ACCESS-NRI/access-om2-configs ?)

**Comment by dougiesquire** [comment]
@chrisb13 feel free to just push to the `dougiesquire/1deg_jra55_ryf_wombatlite` branch (or submit a PR to it)

**Comment by dougiesquire** [comment]
@chrisb13, I've just hijacked outdated PR #76 and pushed changes to build a prerelease that hopefully helps with testing:
- Prerelease `pr76-26` uses git submodule generic tracers (and `MOM5/src/shared` FMS and git submodule mocsy)
- Prerelease `pr82-7` uses "spackified" generic tracers (and ACCESS "spackified" FMS and mocsy)

Everything else should be the same between the prereleases.

We want to check whether these two prereleases give the same answers and performance. I've updated my ACCESS-OM2 [`1deg_jra55_ryf_wombatlite`](https://github.com/ACCESS-NRI/access-om2-configs/tree/dougiesquire/1deg_jra55_ryf_wombatlite) test config to use the `pr76-26` prerelease. This will hopefully run as is and then [this line](https://github.com/ACCESS-NRI/access-om2-configs/blob/da5c03ee3161932ecd1a129b1846b27f3cd37e55/config.yaml#L21) can be changed to use the `pr82-7` prerelease.

**Comment by chrisb13** [comment]
@dougiesquire, thanks for clarifying. 

I think there's something odd about `pr76-26` as a `payu setup` fails to find the ocean executable..
```bash
[cyb561.gadi-login-05: 1deg_jra55_ryf4_dougie]$ payu setup
laboratory path:  /scratch/p73/cyb561/access-om2
binary path:  /scratch/p73/cyb561/access-om2/bin
input path:  /scratch/p73/cyb561/access-om2/input
work path:  /scratch/p73/cyb561/access-om2/work
archive path:  /scratch/p73/cyb561/access-om2/archive
Metadata and UUID generation is disabled. Experiment name used for archival: 1deg_jra55_ryf4_dougie
payu: Found modules in /opt/Modules/v4.3.0
Loading access-om2/pr76-26
  Loading requirement: cice5/2023.10.19 libaccessom2/2023.10.26 mom5/2024.08.23
Loading input manifest: manifests/input.yaml
Loading restart manifest: manifests/restart.yaml
Loading exe manifest: manifests/exe.yaml
Setting up atmosphere
Setting up ocean
Traceback (most recent call last):
  File "/g/data/vk83/apps/payu/1.1.5/bin/payu", line 10, in <module>
    sys.exit(parse())
  File "/g/data/vk83/apps/payu/1.1.5/lib/python3.10/site-packages/payu/cli.py", line 49, in parse
    run_cmd(**args)
  File "/g/data/vk83/apps/payu/1.1.5/lib/python3.10/site-packages/payu/subcommands/setup_cmd.py", line 28, in runcmd
    expt.setup(force_archive=force_archive)
  File "/g/data/vk83/apps/payu/1.1.5/lib/python3.10/site-packages/payu/experiment.py", line 439, in setup
    model.setup()
  File "/g/data/vk83/apps/payu/1.1.5/lib/python3.10/site-packages/payu/models/mom.py", line 66, in setup
    super(Mom, self).setup()
  File "/g/data/vk83/apps/payu/1.1.5/lib/python3.10/site-packages/payu/models/model.py", line 316, in setup
    raise FileNotFoundError(
FileNotFoundError: Executable for ocean model not found on path: /scratch/p73/cyb561/access-om2/bin/fms_ACCESS-OM.x
```

Working backwards our two commits ([yours](https://github.com/ACCESS-NRI/access-om2-configs/commit/da5c03ee3161932ecd1a129b1846b27f3cd37e55) and [mine](https://github.com/chrisb13/access-om2-configs/commit/e41c923a600dbb309d3286803c36976a61f73501)) are very similar with changes only at the config.yaml (both our commits have the same parent):
image
Swapping `pr76-26` for `pr82-7` removes the above error.

Have you been able to run `pr76-26`?

**Comment by anton-seaice** [comment]
For reasons that are not clear to me there is an fms build named `fms_ACCESS-ESM.x` in the `access-om2/pr76-26` module. So it looks like the binary might be there with the wrong name ? Or the wrong [variant](https://github.com/ACCESS-NRI/spack-packages/blob/a49002968d99b0161ba57be15a0dba0f76bdbdbc/packages/mom5/package.py#L28) was built ?

**Comment by dougiesquire** [comment]
> Have you been able to run pr76-26?

@chrisb13 I hadn't but trying just now I get the same issue.

Something has gone wrong with the environment module for `pr76-26` and it includes MOM binaries from an ACCESS-ESM1.6 prerelease (??). I've [pinged the release team for help in pr76](https://github.com/ACCESS-NRI/ACCESS-OM2/pull/76#issuecomment-2434803318).

**Comment by CodeGat** [comment]
@harshula note the errors with rebasing on Model Deployment Repositories. The environment `pr82-9` will not reflect the changes accurately.

**Comment by CodeGat** [comment]
Hey all, due to:
* [The update to Prerelease versioning](https://github.com/ACCESS-NRI/build-cd/pull/167) (in which we now use number of deployments rather than number of commits, which allows rebasing), and
* [The update to Projections in the `spack.yaml`](https://github.com/ACCESS-NRI/ACCESS-OM2/issues/84) 

Some of the `spack` environment names may clash for future deployments in this PR. If possible, can we move this to a new PR? Happy to help out if needed.

**Pull request closed without merging**